File: frontend/src/api/departments.ts
import { api } from "./client";
import type { Department, DepartmentInput } from "../types";

export async function listDepartments(): Promise<Department[]> {
  const { data } = await api.get<Department[]>("/departments");
  return data;
}

export async function createDepartment(
  input: DepartmentInput,
): Promise<Department> {
  const { data } = await api.post<Department>("/departments", input);
  return data;
}

File: backend/app/routes/departments.py
"""Department routes."""
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.auth.dependencies import get_current_user, require_roles
from app.database import get_db
from app.models.user import UserRole
from app.schemas.department import DepartmentCreate, DepartmentOut
from app.services import department_service

router = APIRouter(prefix="/api/departments", tags=["departments"])


@router.get("", response_model=list[DepartmentOut])
def list_departments(
    db: Session = Depends(get_db),
    _=Depends(get_current_user),
) -> list:
    return department_service.list_departments(db)


@router.post("", response_model=DepartmentOut, status_code=status.HTTP_201_CREATED)
def create_department(
    payload: DepartmentCreate,
    db: Session = Depends(get_db),
    _=Depends(require_roles(UserRole.ADMIN, UserRole.MANAGER)),
):
    return department_service.create_department(db, payload)
